### PR TITLE
Allow creating `CommandContext` with a target entity version

### PR DESCRIPTION
**Pull request by alexander-aleksandrov**
We need to:
1. Extend `Commands.createContext()` to accept parameter `targetVersion`.
2. Have new method `CommandFactory.createContext(int version)`.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@alexander-aleksandrov please resolve conflicts.

File: client/src/main/java/org/spine3/base/Commands.java [hunk 1]
```diff
@@ -95,6 +95,38 @@ public static CommandId generateId() {
...
+                                               UserId userId,
+                                               ZoneOffset zoneOffset,
+                                               int targetVersion) {
+        final CommandContext.Builder result = getContextBuilder(tenantId, userId, zoneOffset);
+        result.setTargetVersion(targetVersion);
+
+        return result.build();
+    }
+
+    private static CommandContext.Builder getContextBuilder(@Nullable TenantId tenantId,
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/base/Commands.java]: `newContextBuilder()`.

File: client/src/main/java/org/spine3/base/Commands.java [hunk 2]
```diff
@@ -298,7 +330,9 @@ public static Command setSchedule(Command command, Duration delay, Timestamp sch
     public static boolean sameActorAndTenant(CommandContext c1, CommandContext c2) {
         checkNotNull(c1);
         checkNotNull(c2);
-        return  c1.getActor().equals(c2.getActor()) &&
-                c1.getTenantId().equals(c2.getTenantId());
+        return c1.getActor()
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/base/Commands.java]: Please test this. It's only partially tested.

File: client/src/main/java/org/spine3/client/CommandFactory.java [hunk 3]
```diff
@@ -108,6 +108,32 @@ public Command create(Message message) {
...
+
+        final CommandContext context = createCommandContext(targetVersion);
+        final Command result = createCommand(message, context);
+        return result;
+    }
+
+    /**
+     * Creates command context for a new command with entity ID.
+     */
+    protected CommandContext createCommandContext(int targetVersion) {
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/client/CommandFactory.java]: Please rename to `createContext()`.

File: client/src/main/java/org/spine3/base/Commands.java (hunk 1 above)
Comment (alexander-aleksandrov) [inline client/src/main/java/org/spine3/base/Commands.java]: Done

File: client/src/main/java/org/spine3/client/CommandFactory.java (hunk 3 above)
Comment (alexander-aleksandrov) [inline client/src/main/java/org/spine3/client/CommandFactory.java]: Renamed

**Comment by alexander-aleksandrov** [comment]
@armiol PTAL

**Comment by armiol** [comment]
@alexander-aleksandrov something is wrong with the PR description.

File: client/src/main/java/org/spine3/base/Commands.java [hunk 4]
```diff
@@ -97,6 +97,41 @@ public static CommandContext createContext(@Nullable TenantId tenantId,
...
+
+    /**
+     * Creates a new command context with the current time.
+     *
+     * <p>This method is not supposed to be called from outside the framework.
+     * Commands in client applications should be created by {@link org.spine3.client.CommandFactory#createCommand(Message)},
+     * which creates {@code CommandContext} automatically.
+     *
+     * @param tenantId      the ID of the tenant or {@code null} for single-tenant applications
+     * @param userId        the actor id
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Commands.java]: `the actor ID`.

File: server/src/main/java/org/spine3/server/aggregate/Aggregate.java [hunk 5]
```diff
@@ -57,7 +57,7 @@
  *
  * <h2>Creating an aggregate class</h2>
  *
- * <p>In order to create a new aggregate class you need to:
+ * <p>In order to createCommand a new aggregate class you need to:
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/Aggregate.java]: Is this an automatic replacement? The text does not make any sense now. It's everywhere in the docs now. 

Please fix.

File: client/src/main/java/org/spine3/base/Commands.java (hunk 4 above)
Comment (alexander-aleksandrov) [inline client/src/main/java/org/spine3/base/Commands.java]: Fixed

File: server/src/main/java/org/spine3/server/aggregate/Aggregate.java (hunk 5 above)
Comment (alexander-aleksandrov) [inline server/src/main/java/org/spine3/server/aggregate/Aggregate.java]: Yes probably I did auto replacement. 
Fixed.

**Comment by alexander-aleksandrov** [comment]
@armiol PTAL

File: client/src/test/java/org/spine3/base/CommandsShould.java [hunk 6]
```diff
@@ -64,6 +69,11 @@
     private final StringValue stringValue = newStringValue(newUuid());
 
     @Test
+    public void have_private_ctor() {
```
Comment (armiol) [inline client/src/test/java/org/spine3/base/CommandsShould.java]: Could you please add one more test using the `NullToleranceTest` utility to ensure the `Commands` methods aren't vulnerable?

**Comment by alexander-aleksandrov** [comment]
@armiol PTAL

File: client/src/test/java/org/spine3/base/CommandsShould.java (hunk 6 above)
Comment (alexander-aleksandrov) [inline client/src/test/java/org/spine3/base/CommandsShould.java]: Done

File: client/src/test/java/org/spine3/base/CommandsShould.java [hunk 7]
```diff
@@ -74,6 +76,19 @@ public void have_private_ctor() {
     }
 
     @Test
+    public void pass_the_null_tolerance_check() throws Descriptors.DescriptorValidationException {
+        final NullToleranceTest nullToleranceTest =
+                NullToleranceTest.newBuilder()
+                                 .setClass(Commands.class)
+                                 .addDefaultValue(FileDescriptor.buildFrom(
```
Comment (armiol) [inline client/src/test/java/org/spine3/base/CommandsShould.java]: Please extract the default value as a variable.

**Comment by alexander-yevsyukov** [comment]
@alexander-aleksandrov, please resolve conflicts. Let's finish with this PR.

File: client/src/main/java/org/spine3/base/Commands.java [hunk 8]
```diff
@@ -98,6 +98,42 @@ public static CommandContext createContext(@Nullable TenantId tenantId,
...
 
+        final CommandContext.Builder result = newContextBuilder(tenantId, userId, zoneOffset);
+        return result.build();
+    }
+
+    /**
+     * Creates a new command context with the current time.
+     *
+     * <p>This method is not supposed to be called from outside the framework.
+     * Commands in client applications should be created by {@link org.spine3.client.CommandFactory#createCommand(Message)},
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/base/Commands.java]: Please have link text after FQN link.

**Pull request merged**